pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -618,9 +607,15 @@ private BrokerResponseNative handleRequest(long requestId, String query,
       }
 
       if (offlineBrokerRequest == null && realtimeBrokerRequest == null) {
-        LOGGER.info("No server found for request {}: {}", requestId, query);
-        _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.NO_SERVER_FOUND_EXCEPTIONS, 1);
-        return new BrokerResponseNative(exceptions);
+        if (!exceptions.isEmpty()) {
+          LOGGER.info("No server found for request {}: {}", requestId, query);
+          _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.NO_SERVER_FOUND_EXCEPTIONS, 1);
+          return new BrokerResponseNative(exceptions);
+        } else {
+          // When all segments have been pruned, we can just return an empty response.
+          return getEmptyBrokerOnlyResponse(requestId, query, requesterIdentity, requestContext, pinotQuery,
```

Comment:
we have `setNumSegmentsPrunedByBroker`, but we haven't been setting that here. I imagine we could, but I feel like that might be a bigger semantic change.